pkg/providers.bzl

```diff
@@ -29,3 +28,69 @@ PackageVariablesInfo = provider(
         "values": "Dict of name/value pairs",
     },
 )
+
+# For the below, attributes always look something like this:
+#
+# ```
+#   {"unix": ["0755", "root", "root"]}
+# ```
+#
```

Comment:
I think it is important to be clear hear that destination is always an absolute path in the final archive, and that rules creating pkg_* do not prepend in the package path. That makes the arithmetic clear.  if you manifest is

== foo/bar/BUILD
manifest = [
  ":my_bin",    "usr/bin/my_bin"
  "glob(["data/***",  "/usr/share/my_bin/lib"
]

that the dest paths will be   /usr/bin/my_bin, /usr/share/my_bin/lib/data/...
And that if you put it in pkg_rpm, the files will be installed exactly there.
Eventually, pkg_filegroup wil have ways to reroot the entities.

Reply:
To confirm, the rules that make the non `PackageFileGroupInfo` providers should not allow the construction of new paths?  That sounds good to me.

Does this affect subtraction at all, a la `strip_prefix`?

Reply:
I'm not sure what you mean by should not allow the construction of new paths.

We have to think about what we want for strip_prefix.  The only thing I know
for certain is that it should apply to incoming deps and change the final
destination paths in the output. But, there is so much room for interpretation
in that. For example
- what happens if all the files in the deps do not have the same prefix? Is that an error, or do only some files strip the prefix?
- do we even need strip_prefix at all? Isn't that just a degenerate case of regex path rewriting?

Reply:
> I'm not sure what you mean by "should not allow the construction of new paths".

Now that I think about it, that was perhaps poorly worded.

I wonder, though, is the overall question rule implementation detaisl?  AFAICT all these providers need to do is inform package builders of how packages should be structured.  Any sort of path reconstruction metadata seems to be out-of-scope for a `Provider` interface.

---

> We have to think about what we want for strip_prefix. The only thing I know
for certain is that it should apply to incoming deps and change the final
destination paths in the output.

Agreed

> - what happens if all the files in the deps do not have the same prefix? Is that an error, or do only some files strip the prefix?
 
The existing `strip_prefix` behavior in the `pkg_filegroup` ignores prefixes that don't match.  I wouldn't mind changing that behavior to failure.

> - do we even need strip_prefix at all? Isn't that just a degenerate case of regex path rewriting?

It is, but we cannot use regexps in starlark directly.  We would have to constrain the syntax significantly and then write a parser, and then that just devolves into a mess.

The other option would be to defer the path reconstruction to the packaging step.  I wouldn't be particularly opposed to this, but since the regexp processing would be handled by external tooling, this could lead to additional confusion.

I feel that regexp substitution would would only complicate matters and should not be added in.

Reply:
Sounds good
- strip_prefix with a non-match is a fail.
- no regex

With those simplifications, we could defer remappings until actual packaging. By that I mean if we had
```
pkg_filegroup(  strip=x, prefix=y, srcs = pkg_filegroup(srcs=[...], prefix=x))
```
we do not have to actually modify the list of paths from the inner group. We could just pass the right things up so that the final packager rule can apply both transforms in order. 

I don't think we have to settle that today.  Let's get out a version that can do the basic mkdir and symlink (which has pent up feature requests) and when we try to implement nice path remappings, we can refine.

Reply:
Agreed.

Reply:
Agreed.

> I think we should talk about overall plan.
> 
> I'm happy enough with most of this to use it to start implementing a few simple things.
> If we had mkdir and symlink, I would slip those into pkg_tar and zip just to prove them out.
> I find that actually building the code shakes out the design issues.

Sounds good to me.  I'll undraftify it now.

The only real outstanding question for me is the singular/plural thing, but we can do that in an additional review as we try this out, if you'd like.